File: tests/human/x/fortran/group_by_multi_join.f90
program group_by_multi_join
  implicit none
  type :: Nation
    integer :: id
    character(len=1) :: name
  end type Nation
  type :: Supplier
    integer :: id
    integer :: nation
  end type Supplier
  type :: PartSupp
    integer :: part
    integer :: supplier
    real :: cost
    integer :: qty
  end type PartSupp
  type(Nation) :: nations(2)
  type(Supplier) :: suppliers(2)
  type(PartSupp) :: partsupp(3)
  real :: total100, total200
  integer :: i,j,k

  nations(1) = Nation(1,'A')
  nations(2) = Nation(2,'B')

  suppliers(1) = Supplier(1,1)
  suppliers(2) = Supplier(2,2)

  partsupp(1) = PartSupp(100,1,10.0,2)
  partsupp(2) = PartSupp(100,2,20.0,1)
  partsupp(3) = PartSupp(200,1,5.0,3)

  total100 = 0
  total200 = 0
  do i = 1,3
    do j = 1,2
      if (partsupp(i)%supplier == suppliers(j)%id) then
        do k = 1,2
          if (suppliers(j)%nation == nations(k)%id .and. nations(k)%name == 'A') then
            if (partsupp(i)%part == 100) total100 = total100 + partsupp(i)%cost*partsupp(i)%qty
            if (partsupp(i)%part == 200) total200 = total200 + partsupp(i)%cost*partsupp(i)%qty
          end if
        end do
      end if
    end do
  end do

  write(*,'("map[part:100 total:",F0.0,"] map[part:200 total:",F0.0,"]")') total100, total200
end program group_by_multi_join

File: archived/tests/transpiler/x/pas/group_by_multi_join.pas
{$mode objfpc}
program Main;
type Anon1 = record
  id: integer;
  name: string;
end;
type Anon2 = record
  id: integer;
  nation: integer;
end;
type Anon3 = record
  part: integer;
  supplier: integer;
  cost: real;
  qty: integer;
end;
type Anon4 = record
  part: integer;
  value: real;
end;
type Anon5 = record
  part: integer;
  total: real;
end;
var
  nations: array of Anon1;
  suppliers: array of Anon2;
  partsupp: array of Anon3;
  filtered: array of Anon4;
  grp4: array of Anon5;
  idx5: integer;
  i6: integer;
  grouped: array of Anon5;
  ps: Anon3;
  s: Anon2;
  n: Anon1;
  x: Anon4;
begin
  nations := [(id: 1; name: 'A'), (id: 2; name: 'B')];
  suppliers := [(id: 1; nation: 1), (id: 2; nation: 2)];
  partsupp := [(part: 100; supplier: 1; cost: 10; qty: 2), (part: 100; supplier: 2; cost: 20; qty: 1), (part: 200; supplier: 1; cost: 5; qty: 3)];
  filtered := [];
  for ps in partsupp do begin
  for s in suppliers do begin
  for n in nations do begin
  if ((s.id = ps.supplier) and (n.id = s.nation)) and (n.name = 'A') then begin
  filtered := concat(filtered, [(part: ps.part; value: ps.cost * ps.qty)]);
end;
end;
end;
end;
  grp4 := [];
  for x in filtered do begin
  idx5 := -1;
  for i6 := 0 to (Length(grp4) - 1) do begin
  if grp4[i6].part = x.part then begin
  idx5 := i6;
  break;
end;
end;
  if idx5 = -1 then begin
  grp4 := concat(grp4, [(part: x.part; total: x.value)]);
end else begin
  grp4[idx5].total := grp4[idx5].total + x.value;
end;
end;
  grouped := grp4;
  writeln(grouped);
end.
